File: src/bridge/moaBridge.ts
/**
 * Android WebView 桥接接口
 * 遵循项目已有的 moaBridge 模式
 * 在 Android 壳中，window.moaBridge 由原生注入
 */

declare global {
  interface Window {
    moaBridge?: MoaBridgeInstance;
    [key: string]: unknown;
  }
}

interface MoaBridgeInstance {
  callNative: (method: string, paramsJson?: string) => void;
  [key: string]: unknown;
}

function callNative(method: string, params?: Record<string, unknown>) {
  const paramsJson = params ? JSON.stringify(params) : undefined;
  window.moaBridge?.callNative(method, paramsJson);
}

// ==================== STT / TTS ====================

interface STTResult {
  success: boolean;
  text?: string;
  message?: string;
}

export const startSTT = async (params: {
  lang?: string;
  cbFuncName?: string;
} = {}): Promise<STTResult> => {
  const cbFuncName = params.cbFuncName ?? 'onSTTResult';
  return new Promise((resolve) => {
    window[cbFuncName] = (data: unknown) => {
      if (data === undefined) return;
      const parsed = typeof data === 'string' ? JSON.parse(data) : data;
      resolve(parsed as STTResult);
      delete window[cbFuncName];
    };
    callNative('startSTT', {
      lang: params.lang ?? 'zh-CN',
      cbFuncName,
    });
  });
};

export const stopSTT = (): void => {
  callNative('stopSTT');
};

interface TTSParams {
  text: string;
  lang?: string;
  rate?: number;
}

export const playTTS = (params: TTSParams): void => {
  callNative('playTTS', {
    text: params.text,
    lang: params.lang ?? 'zh-CN',
    rate: params.rate ?? 1.0,
  });
};

export const stopTTS = (): void => {
  callNative('stopTTS');
};

// ==================== File ====================

interface FilePickResult {
  success: boolean;
  uri?: string;
  name?: string;
  mimeType?: string;
  content?: string;
  isImage?: boolean;
  dataUrl?: string;
  message?: string;
}

export const pickFile = async (params: {
  accept?: string;
  cbFuncName?: string;
} = {}): Promise<FilePickResult> => {
  const cbFuncName = params.cbFuncName ?? 'onFilePicked';
  return new Promise((resolve) => {
    window[cbFuncName] = (data: unknown) => {
      if (data === undefined) return;
      const parsed = typeof data === 'string' ? JSON.parse(data) : data;
      resolve(parsed as FilePickResult);
      delete window[cbFuncName];
    };
    callNative('pickFile', {
      accept: params.accept ?? '*/*',
      cbFuncName,
    });
  });
};

// ==================== Navigation ====================

export const goBack = (): void => {
  callNative('goBack');
};

export const setTitle = (title: string): void => {
  callNative('setTitle', { title });
};

// ==================== SafeArea ====================

interface SafeAreaResult {
  success: boolean;
  top?: number;
  bottom?: number;
}

export const getSafeArea = async (params: {
  cbFuncName?: string;
} = {}): Promise<SafeAreaResult> => {
  const cbFuncName = params.cbFuncName ?? 'onSafeAreaResult';
  return new Promise((resolve) => {
    window[cbFuncName] = (data: unknown) => {
      if (data === undefined) return;
      const parsed = typeof data === 'string' ? JSON.parse(data) : data;
      resolve(parsed as SafeAreaResult);
      delete window[cbFuncName];
    };
    callNative('getSafeArea', { cbFuncName });
  });
};

// ==================== Utils ====================

export const isBridgeAvailable = (): boolean => {
  return typeof window.moaBridge?.callNative === 'function';
};

File: android/app/src/main/java/com/myopenweb/app/bridge/MoaBridge.kt
package com.myopenweb.app.bridge

import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.myopenweb.app.MainActivity
import com.myopenweb.app.native_.NativeFilePicker
import com.myopenweb.app.native_.NativeSTT
import com.myopenweb.app.native_.NativeTTS
import com.myopenweb.app.native_.SafeAreaHelper
import com.myopenweb.app.util.JsonUtil

class MoaBridge(
    private val activity: MainActivity,
    private val webView: WebView
) {

    companion object {
        private const val TAG = "MoaBridge"
    }

    val nativeSTT = NativeSTT(activity, this)
    val nativeTTS = NativeTTS(activity)
    val nativeFilePicker = NativeFilePicker(activity, this)
    val safeAreaHelper = SafeAreaHelper(activity)

    @JavascriptInterface
    fun callNative(method: String, paramsJson: String?) {
        Log.d(TAG, "callNative: $method, params: $paramsJson")

        val params = JsonUtil.parse(paramsJson)
        val cbFuncName = params.optString("cbFuncName", "")

        try {
            when (method) {
                "startSTT" -> {
                    val lang = params.optString("lang", "zh-CN")
                    nativeSTT.start(lang, cbFuncName)
                }
                "stopSTT" -> nativeSTT.stop()

                "playTTS" -> {
                    val text = params.optString("text", "")
                    val lang = params.optString("lang", "zh-CN")
                    val rate = params.optDouble("rate", 1.0).toFloat()
                    nativeTTS.play(text, lang, rate)
                }
                "stopTTS" -> nativeTTS.stop()

                "pickFile" -> {
                    val accept = params.optString("accept", "*/*")
                    nativeFilePicker.pick(accept, cbFuncName)
                }

                "goBack" -> activity.runOnUiThread {
                    @Suppress("DEPRECATION")
                    activity.onBackPressed()
                }

                "setTitle" -> {
                    val title = params.optString("title", "")
                    activity.runOnUiThread { activity.title = title }
                }

                "getSafeArea" -> {
                    val result = safeAreaHelper.getSafeArea()
                    invokeCallback(cbFuncName, result)
                }

                else -> Log.w(TAG, "Unknown method: $method")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error in callNative($method)", e)
            if (cbFuncName.isNotEmpty()) {
                invokeCallback(cbFuncName, JsonUtil.error(e.message ?: "Unknown error"))
            }
        }
    }

    fun invokeCallback(cbFuncName: String, jsonData: String) {
        if (cbFuncName.isEmpty()) return
        val escaped = jsonData.replace("\\", "\\\\").replace("'", "\\'")
        val script = "if(window['$cbFuncName']){window['$cbFuncName']('$escaped')}"
        activity.runOnUiThread {
            webView.evaluateJavascript(script, null)
        }
    }

    fun destroy() {
        nativeSTT.destroy()
        nativeTTS.destroy()
    }
}
